Calcite 5418 3

jamesstarr opened the pull request:
Rework RexSubquery to contain the Correlate Id and for the correlate id to be used in subquery expansions.  This fixes bugs with nested correlated subqueries when expand is false.

zinking commented on core/src/main/java/org/apache/calcite/rel/logical/LogicalFilter.java:
```diff
@@ -47,7 +47,7 @@
  */
 public final class LogicalFilter extends Filter {
-  private final ImmutableSet<CorrelationId> variablesSet;
+  private final Set<CorrelationId> variablesSet;
```
is it possible to keep the immutableSet here? generally Immutable should be preferred.

jamesstarr replied:
Is there any other changes you would like me to make for this?  The set is currently always immutable, just the signature of the member variable is not.

mihaibudiu commented:
Can we revive this PR?

jamesstarr commented:
@mihaibudiu
> Can we revive this PR?
I will look at rebasing it next week if I can get a review.

mihaibudiu commented:
I don't know this part of the code that well, but I will try to supply a review.
This PR is also quite large.
But it looks like it's a bug worth fixing.

jamesstarr committed:
[CALCITE-5418] Rework Correlated Id Subquery Expansion

Remorking RexSubquery to contain the correlate id for resolving applicable variables.  This
simplifies probigating the corralate id to the subquery removal.

Changes:
* Adding Utils.getOnlyElement do to guava not maintianing the nullability
* Adding CorrelationId to RexSubquery
* Using Set instead of ImmutableSet for LogicalFitler.variableSet
* Helper methods in RexUtil for finding CorrelateIds in RexNode
* SqlValidator explicitly pass in a lateral scope instead of deriving it from using scope
* Renaming TableScope to LateralTableScope
* Fixing isWithin semantic so a LateralScope.isWithIn(JoinScope) not vica versa
* Reworking SqlValidatorImpl to directly expand subqueries instead of using a map
* In SubQueryRemoveRule using CorrelateId dervivied from SubQueries instead of references of CorrelateId
